core/chains/evm/headtracker/head_listener.go

```diff
@@ -28,32 +32,68 @@ var (
 	}, []string{"evmChainID"})
 )
 
-type headListener struct {
-	config           Config
-	ethClient        evmclient.Client
+type headListener[
+	H commontypes.Head[BLOCK_HASH],
+	HTH htrktypes.Head[H, BLOCK_HASH, ID],
+	S commontypes.Subscription,
+	ID txmgrtypes.ID,
+	BLOCK_HASH commontypes.Hashable,
+	CLIENT htrktypes.Client[HTH, S, ID, BLOCK_HASH],
+] struct {
+	config           htrktypes.Config
+	client           CLIENT
 	logger           logger.Logger
 	chStop           utils.StopChan
-	chHeaders        chan *evmtypes.Head
-	headSubscription ethereum.Subscription
+	chHeaders        chan HTH
+	headSubscription commontypes.Subscription
 	connected        atomic.Bool
 	receivingHeads   atomic.Bool
+	getNilHead       func() H
 }
 
-// NewHeadListener creates a new HeadListener
-func NewHeadListener(lggr logger.Logger, ethClient evmclient.Client, config Config, chStop chan struct{}) httypes.HeadListener {
-	return &headListener{
-		config:    config,
-		ethClient: ethClient,
-		logger:    lggr.Named("HeadListener"),
-		chStop:    chStop,
+type evmHeadListener = headListener[*evmtypes.Head, *evmtypes.Head, ethereum.Subscription, *big.Int, common.Hash, evmclient.Client]
+
+func NewHeadListener[
+	H commontypes.Head[BLOCK_HASH],
+	HTH htrktypes.Head[H, BLOCK_HASH, ID],
+	S commontypes.Subscription,
+	ID txmgrtypes.ID,
+	BLOCK_HASH commontypes.Hashable,
+	CLIENT htrktypes.Client[HTH, S, ID, BLOCK_HASH],
+](
+	lggr logger.Logger,
+	client CLIENT,
+	config Config, chStop chan struct{},
+	getNilHead func() H,
+) *headListener[H, HTH, S, ID, BLOCK_HASH, CLIENT] {
+	return &headListener[H, HTH, S, ID, BLOCK_HASH, CLIENT]{
+		config:     NewWrappedConfig(config),
+		client:     client,
+		logger:     lggr.Named("HeadListener"),
+		chStop:     chStop,
+		getNilHead: getNilHead,
 	}
 }
 
-func (hl *headListener) Name() string {
+func NewEvmHeadListener(
+	lggr logger.Logger,
+	ethClient evmclient.Client,
+	config Config, chStop chan struct{},
+) *evmHeadListener {
+	return NewHeadListener[*evmtypes.Head, *evmtypes.Head,
+		ethereum.Subscription, *big.Int, common.Hash,
+	](lggr, ethClient, config, chStop,
```

Comment:
Do we have to list out these types explicitly? Or can they be inferred from the arguments?
```suggestion
	return NewHeadListener(lggr, ethClient, config, chStop,
```

Reply:
Yup, i do think we would have to list them out explicitly. Similar to the case of `NewTxm`

https://github.com/smartcontractkit/chainlink/blob/209324b3a52b3958a8798c94c91554efae2a54e6/core/chains/evm/txmgr/txmgr.go#L132